server/src/main/java/org/elasticsearch/search/aggregations/pipeline/PercentileRanksBucketPipelineAggregationBuilder.java

```diff
@@ -0,0 +1,182 @@
+        }
+
+        this.values = values;
+        return this;
+    }
+
+    /**
+     * Set whether the XContent should be keyed
+     */
+    public PercentileRanksBucketPipelineAggregationBuilder setKeyed(boolean keyed) {
+        this.keyed = keyed;
+        return this;
+    }
+
+    /**
+     * Get whether the XContent should be keyed
+     */
+    public boolean getKeyed() {
+        return keyed;
+    }
+
+    @Override
+    protected PipelineAggregator createInternal(Map<String, Object> metaData) {
+        return new PercentileRanksBucketPipelineAggregator(name, values, keyed, bucketsPaths, gapPolicy(), formatter(), metaData);
+    }
+
+    @Override
+    public void doValidate(AggregatorFactory<?> parent, Collection<AggregationBuilder> aggFactories,
+                           Collection<PipelineAggregationBuilder> pipelineAggregatorFactories) {
+        super.doValidate(parent, aggFactories, pipelineAggregatorFactories);
+    }
+
+    @Override
+    protected XContentBuilder doXContentBody(XContentBuilder builder, Params params) throws IOException {
+        if (values != null) {
+            builder.array(VALUES_FIELD.getPreferredName(), values);
+        }
+        builder.field(KEYED_FIELD.getPreferredName(), keyed);
+        return builder;
+    }
+
+    public static final PipelineAggregator.Parser PARSER = new BucketMetricsParser() {
+
+        @Override
+        protected PercentileRanksBucketPipelineAggregationBuilder buildFactory(String pipelineAggregatorName,
+                                                                           String bucketsPath, Map<String, Object> params) {
+
+            PercentileRanksBucketPipelineAggregationBuilder factory = new
+                PercentileRanksBucketPipelineAggregationBuilder(pipelineAggregatorName, bucketsPath);
+
+            double[] values = (double[]) params.get(VALUES_FIELD.getPreferredName());
+            if (values != null) {
+                factory.setvalues(values);
+            }
+            Boolean keyed = (Boolean) params.get(KEYED_FIELD.getPreferredName());
+            if (keyed != null) {
+                factory.setKeyed(keyed);
+            }
+
+            return factory;
+        }
+
+        @Override
+        protected boolean token(XContentParser parser, String field, XContentParser.Token token, Map<String, Object> params)
+            throws IOException {
+            if (VALUES_FIELD.match(field, parser.getDeprecationHandler()) && token == XContentParser.Token.START_ARRAY) {
+                DoubleArrayList values = new DoubleArrayList(10);
+                while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+                    values.add(parser.doubleValue());
+                }
+                params.put(VALUES_FIELD.getPreferredName(), values.toArray());
+                return true;
+            }
+            else if (KEYED_FIELD.match(field, parser.getDeprecationHandler()) && token == XContentParser.Token.VALUE_BOOLEAN){
```

Comment:
Style nit: let's move this `else if` up a line so it comes after the curly bracket.

Reply:
got it!